pkg/upgrade/v1alpha2/kubernetes.go

```diff
@@ -0,0 +1,442 @@
+}
+
+// Get returns a upgrade task object
+func (k *Kubeclient) Get(
+	name string,
+	opts metav1.GetOptions,
+) (*apis.UpgradeTask, error) {
+	if strings.TrimSpace(name) == "" {
+		return nil, errors.New(
+			"failed to get upgradeTask: missing upgradeTask name",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get upgradeTask {%s} in namespace {%s}",
+			name,
+			k.namespace,
+		)
+	}
+	return k.get(cli, k.namespace, name, opts)
+}
+
+// Create creates a upgradeTask in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error) {
+	if upgradeTask == nil {
+		return nil, errors.New(
+			"failed to create upgradeTask: nil upgradeTask object",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to create upgradeTask {%s} in namespace {%s}",
+			upgradeTask.Name,
+			upgradeTask.Namespace,
+		)
+	}
+	return k.create(cli, k.namespace, upgradeTask)
+}
+
+// Delete deletes a upgradeTask instance from the
+// kubecrnetes cluster
+func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error {
+	if strings.TrimSpace(name) == "" {
+		return errors.New("failed to delete upgradeTask: missing upgradeTask name")
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"failed to delete upgradeTask {%s} in namespace {%s}",
+			name,
+			k.namespace,
+		)
+	}
+	return k.del(cli, k.namespace, name, deleteOpts)
+}
+
+// Patch patches the upgrade task  if present in kubernetes cluster
+func (k *Kubeclient) Patch(
+	name string,
+	pt types.PatchType,
+	data []byte,
+	subresources ...string,
+) (*apis.UpgradeTask, error) {
+	if len(name) == 0 {
+		return nil, errors.New(
+			"failed to patch upgrade task : missing upgradeTask name",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to patch upgradeTask: {%s}", name)
+	}
+	return k.patch(cli, k.namespace, name, pt, data, subresources...)
+}
+
+// Update updates the upgrade task  if present in kubernetes cluster
+func (k *Kubeclient) Update(
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error) {
+	if upgradeTask == nil {
+		return nil, errors.New(
+			"failed to udpate upgradeTask: nil upgradeTask object",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to update upgradeTask {%s} in namespace {%s}",
+			upgradeTask.Name,
+			upgradeTask.Namespace,
+		)
+	}
+	return k.update(cli, k.namespace, upgradeTask)
+}
+
+// GetUpgradeDetailedStatuses builds and returns the UpgradeDetailedStatuses
+// object with given parameters
+func GetUpgradeDetailedStatuses(
+	step apis.UpgradeStep,
+	phase apis.StepPhase,
+	// args contains message and reason depending upon the StepPhase
+	args ...string,
```

Comment:
Should we have step, phase & all args as `StatusOptions`
e.g.
```go
type StatusOption struct {
  Step string
  Phase string
  Message string
  Reason string
}
```
```go
func IsValidStatus(o StatusOption) bool {
  if o.Step == "" {...}
  if o.Phase == "" {...}
  if o.Message == "" && o.Phase != WAITING {...}
  if o.Reason == "" && o.Phase != COMPLETED {...}
}
```

Reply:
Instead of creating the new struct can we pass the original struct for validations only and don't build the struct here at all.
```go
// IsValidStatus is used to validate UpgradeDetailedStatuses
func IsValidStatus(o apis.UpgradeDetailedStatuses) bool {
	if o.Step == "" {
		return false
	}
	if o.Phase == "" {
		return false
	}
	if o.Message == "" && o.Phase != apis.StepWaiting {
		return false
	}
	if o.Reason == "" && o.Phase != apis.StepCompleted {
		return false
	}
	return true
}
```
This is the function we will use to update the CR.
```go
func updateUpgradeDetailedStatus(
	uStatusObj apis.UpgradeDetailedStatuses, openebsNamespace string,
) error {
	if utask.IsValidStatus(uStatusObj) {
		return errors.Errorf(
			"failed to update upgradetask status: invalid status %v",
			uStatusObj,
		)
	}
	uStatusObj.LastUpdatedTime = metav1.Now()
	if phase == apis.StepWaiting {
		uStatusObj.StartTime = uStatusObj.LastUpdatedTime
		utaskObj.Status.UpgradeDetailedStatuses = append(
			utaskObj.Status.UpgradeDetailedStatuses,
			uStatusObj,
		)
	} else {
		l := len(utaskObj.Status.UpgradeDetailedStatuses)
		uStatusObj.StartTime = utaskObj.Status.UpgradeDetailedStatuses[l-1].StartTime
		utaskObj.Status.UpgradeDetailedStatuses[l-1] = uStatusObj
	}
	utaskObj, err = utaskClient.WithNamespace(openebsNamespace).Update(utaskObj)
	if err != nil {
		return errors.Wrapf(err, "failed to update upgradetask ")
	}
	return nil
}
```
And pass the actual struct like
```go
updateUpgradeDetailedStatus(
		apis.UpgradeDetailedStatuses{
			Step: apis.PreUpgrade,
			Status: apis.Status{
				Phase: apis.StepWaiting,
			},
		},
		openebsNamespace,
	)
```

Reply:
Well.. I am finding it difficult to understand `updateUpgradeDetailedStatus` logic

Reply:
I see a lot of stuff happening in `updateUpgradeDetailedStatus`

Reply:
What is the package that has `updateUpgradeDetailedStatus`? We can determine based on that.